File: tests/codes/fortran/Basic/Basic.f90
module mod_basic

implicit none

    integer :: code_state = 1
    real :: timestamp = 0

    interface codeStep
    module procedure code_step_ids_no_parameters, code_step_only_parameters, code_step_ids_parameters
    module procedure code_step_only_parameters_legacy, code_step_ids_parameters_legacy
    end interface codeStep

    interface initCode
        module procedure init_no_ids_no_parameters, init_only_parameters, init_only_ids, init_ids_parameters
        module procedure init_only_parameters_legacy, init_ids_parameters_legacy
    end interface initCode

    interface cleanUp
        module procedure clean_up_no_ids_no_parameters, clean_up_only_parameters, &
                clean_up_only_ids, clean_up_ids_parameters
        module procedure clean_up_only_parameters_legacy, clean_up_ids_parameters_legacy
    end interface cleanUp

contains
    ! =================================================================================================================
    !                                           CODE STATE SUBROUTINES
    ! =================================================================================================================

    !
    !    RESTORE_CODE_STATE SUBROUTINE
    !
    subroutine restore_code_state (state_str, status_code, status_message)

        implicit none
        character(len=:), allocatable, intent(in) :: state_str
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message


        ! Setting status to SUCCESS
        status_code = 0
        allocate(character(50):: status_message)
        status_message = 'OK'

        read(state_str , *) code_state
        write(*,*) '======================================='
        write(*,*) 'Basic: RESTORE STATE called'
        write(*,*) 'STATE TO BE RESTORED :', code_state
        write(*,*) '======================================='

    end subroutine restore_code_state

    !
    !    GET_CODE_STATE SUBROUTINE
    !
    subroutine get_code_state (state_str, status_code, status_message)

        implicit none
        character(len=:), allocatable, intent(out) :: state_str
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message


        ! Setting status to SUCCESS
        status_code = 0
        allocate(character(50):: status_message)
        status_message = 'OK'

        allocate(character(50):: state_str)
        write(state_str,*) code_state



        write(*,*) '======================================='
        write(*,*) 'Basic: GET CODE STATE called'
        write(*,*) 'STATE is :', state_str
        write(*,*) '======================================='

    end subroutine get_code_state

    ! =================================================================================================================
    !                                           TIMESTAMP SUBROUTINES
    ! =================================================================================================================

    ! - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    !                                   GET TIMESTAMP SUBROUTINE
    ! - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
    subroutine get_timestamp(timestamp_out, error_flag, error_message)

        integer,parameter :: DP=kind(1.0D0)
        real(kind=DP), intent(out) :: timestamp_out
        !----  Status info  ----
        integer, intent(out) :: error_flag
        character(len=:), pointer, intent(out) :: error_message

        timestamp_out = timestamp

    end subroutine get_timestamp

    ! =================================================================================================================
    !                                           INIT SUBROUTINES
    ! =================================================================================================================

    !
    !    INITIALISATION SUBROUTINE
    !
    subroutine init_no_ids_no_parameters(status_code, status_message)
        use ids_schemas, only: ids_parameters_input
        implicit none
        character(:), allocatable :: parameters_string
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        call init_only_parameters (parameters_string, status_code, status_message)

    end subroutine init_no_ids_no_parameters


    !
    !    INITIALISATION SUBROUTINE
    !
    subroutine init_only_parameters_legacy(al_code_parameters, status_code, status_message)
        use ids_schemas, only: ids_parameters_input, ids_core_profiles, ids_distribution_sources
        implicit none
        type (ids_core_profiles) :: core_profiles_in
        type (ids_distribution_sources) :: distribution_sources_out
        type(ids_parameters_input), intent(in) :: al_code_parameters
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message


        call init_ids_parameters_legacy (core_profiles_in, distribution_sources_out, &
                al_code_parameters, status_code, status_message)

    end subroutine init_only_parameters_legacy

    !
    !    INITIALISATION SUBROUTINE
    !
    subroutine init_only_parameters(parameters_string, status_code, status_message)
        use ids_schemas, only: ids_parameters_input, ids_core_profiles, ids_distribution_sources
        implicit none
        type (ids_core_profiles) :: core_profiles_in
        type (ids_distribution_sources) :: distribution_sources_out
        character(:), allocatable :: parameters_string
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message


        call init_ids_parameters (core_profiles_in, distribution_sources_out, &
                parameters_string, status_code, status_message)

    end subroutine init_only_parameters

    !
    !    INITIALISATION SUBROUTINE
    !
    subroutine init_only_ids (core_profiles_in, distribution_sources_out,  status_code, status_message)
        use ids_schemas, only: ids_parameters_input, ids_core_profiles, ids_distribution_sources
        implicit none
        type (ids_core_profiles), intent(in) :: core_profiles_in
        type (ids_distribution_sources), intent(out) :: distribution_sources_out
        character(:), allocatable :: parameters_string
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        call init_ids_parameters (core_profiles_in, distribution_sources_out, &
                parameters_string, status_code, status_message)

    end subroutine init_only_ids

    subroutine init_ids_parameters_legacy (core_profiles_in, distribution_sources_out, &
            al_code_parameters, status_code, status_message)
        use ids_schemas, only: ids_parameters_input, ids_core_profiles, ids_distribution_sources, IDS_STRING_LENGTH
        implicit none
        type (ids_core_profiles), intent(in) :: core_profiles_in
        type (ids_distribution_sources), intent(out) :: distribution_sources_out
        type (ids_parameters_input), intent(in) :: al_code_parameters
        integer, intent(inout) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        character(:), allocatable :: parameters_string
        integer :: string_size

        string_size = IDS_STRING_LENGTH * size(al_code_parameters%parameters_value)

        allocate(character(string_size) :: parameters_string)
        parameters_string = TRANSFER(al_code_parameters%parameters_value, parameters_string)

        call init_ids_parameters (core_profiles_in, distribution_sources_out, &
                parameters_string, status_code, status_message)

    end subroutine init_ids_parameters_legacy
    !
    !    INITIALISATION SUBROUTINE
    !
    subroutine init_ids_parameters (core_profiles_in, distribution_sources_out, &
            parameters_string, status_code, status_message)
        use ids_schemas, only: ids_parameters_input, ids_core_profiles, ids_distribution_sources
        implicit none
        type (ids_core_profiles), intent(in) :: core_profiles_in
        type (ids_distribution_sources), intent(out) :: distribution_sources_out
        character(:), allocatable :: parameters_string
        integer, intent(inout) :: status_code
        character(len=:), pointer, intent(out) :: status_message
        integer :: i


        ! Setting status to SUCCESS
        status_code = 0
        allocate(character(50):: status_message)
        status_message = 'OK'

        write(*,*) '======================================='
        write(*,*) 'Code lifecycle: INITIALISATION called'
        write(*,*) '---------------------------------------'

        distribution_sources_out%ids_properties%homogeneous_time = core_profiles_in%ids_properties%homogeneous_time

        allocate(distribution_sources_out%time(size(core_profiles_in%time)))

        write(0,*) 'Received size of input time from core_profiles_in : ', SIZE(core_profiles_in%time)

        ! Fill in the output IDS (Physical data)
        do i=1,size(core_profiles_in%time)
           ! Time : copy from input IDS
           distribution_sources_out%time(i) = core_profiles_in%time(i)  + 1.0
        enddo

        allocate(distribution_sources_out%code%name(1))   ! For a string of 132 characters max.
        distribution_sources_out%code%name(1)   = 'basic methods: INIT'
        allocate(distribution_sources_out%code%output_flag(size(core_profiles_in%time)))
        distribution_sources_out%code%output_flag(:) = 0

        write(*,*) '---------------------------------------'
        write(*,*) 'Basic methods: INITIALISATION ends'
        write(*,*) '======================================='

    end subroutine init_ids_parameters

    ! =================================================================================================================
    !                                           FINALIZE SUBROUTINES
    ! =================================================================================================================

    !
    !    FINALISATION SUBROUTINE
    !
    subroutine clean_up_ids_parameters(distribution_sources_in, core_profiles_out, &
            parameters_string, status_code, status_message)
        use ids_schemas, only: ids_core_profiles, ids_distribution_sources
        implicit none

        type (ids_distribution_sources), intent(IN) :: distribution_sources_in
        type (ids_core_profiles), intent(OUT) :: core_profiles_out
        character(:), allocatable :: parameters_string

        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message
        integer :: i


        ! Setting status to SUCCESS
        status_code = 0
        allocate(character(50):: status_message)
        status_message = 'OK'

        write(*,*) '======================================='
        write(*,*) 'Basic methods: FINALISATION called'
        write(*,*) '---------------------------------------'


        core_profiles_out%ids_properties%homogeneous_time = distribution_sources_in%ids_properties%homogeneous_time

        allocate(core_profiles_out%time(size(distribution_sources_in%time)))

        write(0,*) 'Received size of input time from core_profiles_in : ', SIZE(distribution_sources_in%time)

        ! Fill in the output IDS (Physical data)
        do i=1,size(distribution_sources_in%time)
           ! Time : copy from input IDS
           core_profiles_out%time(i) = distribution_sources_in%time(i)  + 1.0
        enddo

        allocate(core_profiles_out%code%name(1))   ! For a string of 132 characters max.
        core_profiles_out%code%name(1)   = 'basic methods: FINALIZE'
        allocate(core_profiles_out%code%output_flag(size(distribution_sources_in%time)))
        core_profiles_out%code%output_flag(:) = 0

        write(*,*) '---------------------------------------'
        write(*,*) 'Basic methods: FINALISATION ends'
        write(*,*) '======================================='

    end subroutine clean_up_ids_parameters


    subroutine clean_up_no_ids_no_parameters(status_code, status_message)
        use ids_schemas, only: ids_parameters_input
        implicit none

        character(:), allocatable :: parameters_string
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        call clean_up_only_parameters(parameters_string, status_code, status_message)
    end subroutine clean_up_no_ids_no_parameters


    subroutine clean_up_only_parameters(parameters_string, status_code, status_message)
        use ids_schemas, only: ids_parameters_input, ids_core_profiles, ids_distribution_sources
        implicit none
        type (ids_distribution_sources) :: distribution_sources_in
        type (ids_core_profiles) :: core_profiles_out
        character(:), allocatable :: parameters_string
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message


        call clean_up_ids_parameters (distribution_sources_in, core_profiles_out, &
                parameters_string, status_code, status_message)
    end subroutine clean_up_only_parameters

    subroutine clean_up_only_ids(distribution_sources_in, core_profiles_out, status_code, status_message)
        use ids_schemas, only: ids_core_profiles, ids_distribution_sources
        implicit none

        type (ids_distribution_sources), intent(IN) :: distribution_sources_in
        type (ids_core_profiles), intent(OUT) :: core_profiles_out
        character(:), allocatable :: parameters_string

        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        call clean_up_ids_parameters(distribution_sources_in, core_profiles_out, &
                parameters_string, status_code, status_message)
    end subroutine clean_up_only_ids

    subroutine clean_up_ids_parameters_legacy(distribution_sources_in, core_profiles_out, &
            al_code_parameters, status_code, status_message)
        use ids_schemas, only: ids_core_profiles, ids_distribution_sources, ids_parameters_input, IDS_STRING_LENGTH
        implicit none

        type (ids_distribution_sources), intent(IN) :: distribution_sources_in
        type (ids_core_profiles), intent(OUT) :: core_profiles_out
        type(ids_parameters_input) :: al_code_parameters
        character(:), allocatable :: parameters_string
        integer :: string_size

        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        string_size = IDS_STRING_LENGTH * size(al_code_parameters%parameters_value)

        allocate(character(string_size) :: parameters_string)
        parameters_string = TRANSFER(al_code_parameters%parameters_value, parameters_string)

        call clean_up_ids_parameters(distribution_sources_in, core_profiles_out, &
                parameters_string, status_code, status_message)
    end subroutine clean_up_ids_parameters_legacy

    subroutine clean_up_only_parameters_legacy(al_code_parameters, status_code, status_message)
        use ids_schemas, only: ids_core_profiles, ids_distribution_sources, ids_parameters_input, IDS_STRING_LENGTH
        implicit none

        type (ids_distribution_sources) :: distribution_sources_in
        type (ids_core_profiles) :: core_profiles_out
        type(ids_parameters_input) :: al_code_parameters
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        call clean_up_ids_parameters_legacy(distribution_sources_in, core_profiles_out, &
                al_code_parameters, status_code, status_message)

    end subroutine clean_up_only_parameters_legacy


    ! =================================================================================================================
    !                                           MAIN SUBROUTINES
    ! =================================================================================================================

    !
    !    MAIN SUBROUTINE
    !
    subroutine code_step_no_ids_no_parameters(status_code, status_message)
        use ids_schemas, only: ids_parameters_input
        implicit none

        character(:), allocatable :: parameters_string
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        call code_step_only_parameters(parameters_string, status_code, status_message)
    end subroutine code_step_no_ids_no_parameters

    !
    !    MAIN SUBROUTINE
    !
    subroutine code_step_only_parameters(parameters_string, status_code, status_message)
        use ids_schemas, only: ids_equilibrium, ids_parameters_input, ids_is_valid
        implicit none

        type(ids_equilibrium):: equilibrium_in,equilibrium_out
        character(:), allocatable :: parameters_string
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        call code_step_ids_parameters(equilibrium_in, equilibrium_out, parameters_string, status_code, status_message)
    end subroutine code_step_only_parameters

    subroutine code_step_only_parameters_legacy(al_code_parameters, status_code, status_message)
        use ids_schemas, only: ids_equilibrium, ids_parameters_input, ids_is_valid
        implicit none

        type(ids_equilibrium):: equilibrium_in,equilibrium_out
        type(ids_parameters_input) :: al_code_parameters
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        call code_step_ids_parameters_legacy(equilibrium_in, equilibrium_out, &
                al_code_parameters, status_code, status_message)
    end subroutine code_step_only_parameters_legacy

    !
    !    MAIN SUBROUTINE
    !
    subroutine code_step_ids_no_parameters(equilibrium_in, equilibrium_out, status_code, status_message)
        use ids_schemas, only: ids_equilibrium, ids_parameters_input, ids_is_valid
        use ids_routines, only: ids_copy
        implicit none

        type(ids_equilibrium):: equilibrium_in,equilibrium_out
        character(:), allocatable :: parameters_string
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        call code_step_ids_parameters(equilibrium_in, equilibrium_out, parameters_string, status_code, status_message)
    end subroutine code_step_ids_no_parameters

    !
    !    MAIN SUBROUTINE
    !
    subroutine code_step_ids_parameters(equilibrium_in, equilibrium_out, parameters_string, status_code, status_message)
        use ids_schemas, only: ids_equilibrium, ids_parameters_input, ids_is_valid
        use ids_routines, only: ids_copy
        use mod_parameters_utils

        implicit none

        type(ids_equilibrium):: equilibrium_in,equilibrium_out
        character(:), allocatable :: parameters_string
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        integer :: ntimes = 1
        real :: multiplication_factor = 1

        ! INITIALISATION OF ERROR FLAG
        status_code = 0
        allocate(character(50):: status_message)

        ! READ PARAMETERS
        call extract_values(parameters_string, ntimes, multiplication_factor)

        ! INITIAL DISPLAY
        write(*,*) '======================================='
        write(*,*) 'START OF PHYSICS CODE'
        write(*,*) '---------------------------------------'

        ! CHECK IF INPUT IDS IS VALID
        if (ids_is_valid(equilibrium_in%ids_properties%homogeneous_time).eqv..True. &
            .and.size(equilibrium_in%time)>0) then

            ! COPY THE INPUT IDS IN THE OUTPUT IDS
            call ids_copy(equilibrium_in, equilibrium_out)

            ! MANDATORY FLAG (UNIFORM TIME HERE)
            equilibrium_out%ids_properties%homogeneous_time = 1
            allocate(equilibrium_out%code%name(1))   ! For a string of 132 characters max.

            equilibrium_out%code%name(1)   = 'EXAMPLE: STEP method'
            allocate(equilibrium_out%code%version(1))   ! For a string of 132 characters max.

            ! MODIFY OUTPUT FLAG
            allocate(equilibrium_out%code%output_flag(size(equilibrium_in%time)))
            equilibrium_out%code%output_flag(:) = equilibrium_in%time(1) * &
                ntimes * multiplication_factor * code_state

            status_message = 'STEP: OK'
        else
            ! ERROR IF THE CODE DOES NOT COMPLETE TO THE END
            status_code = -1
            status_message = 'Error in STEP: input IDS not valid'
        endif

        code_state = code_state + 1
        timestamp = timestamp + 10

        ! FINAL DISPLAY
        write(*,*) '---------------------------------------'
        write(*,*) 'END OF PHYSICS CODE'
        write(*,*) '======================================='

    end subroutine code_step_ids_parameters

    !
    !    MAIN SUBROUTINE
    !
    subroutine code_step_ids_parameters_legacy(equilibrium_in, equilibrium_out, &
            al_code_parameters, status_code, status_message)
        use ids_schemas, only: ids_equilibrium, ids_parameters_input, ids_is_valid, IDS_STRING_LENGTH
        use ids_routines, only: ids_copy

        implicit none

        type(ids_equilibrium):: equilibrium_in,equilibrium_out
        type(ids_parameters_input) :: al_code_parameters
        integer, intent(out) :: status_code
        character(len=:), pointer, intent(out) :: status_message

        character(:), allocatable :: parameters_string
        integer :: string_size

        string_size = IDS_STRING_LENGTH * size(al_code_parameters%parameters_value)

        allocate(character(string_size) :: parameters_string)
        parameters_string = TRANSFER(al_code_parameters%parameters_value, parameters_string)

        call code_step_ids_parameters(equilibrium_in, equilibrium_out, parameters_string, status_code, status_message)

    end subroutine code_step_ids_parameters_legacy


end module mod_basic

File: tests/codes/cpp/Basic/Basic.cpp
#if AL_MAJOR == 5
    #include "ALClasses.h"
#elif AL_MAJOR == 4
    #include "UALClasses.h"
#else
    #warning Could not find AL_MAJOR variable. Assuming AL version = 5.x.x.
    #include "ALClasses.h"
#endif

#include "Basic.h"
#include "ParametersUtils.h"
#include "json/json.h"
#include "xmlIMAS.h"
#include <boost/algorithm/string.hpp>
#include <memory>
#include <stdexcept>

// ================================================================================================================
//                                           CODE STATE SUBROUTINES
// ================================================================================================================

// =======================================
//             GET STATE
// =======================================

int code_state = 1;
double timestamp = 0.0;

void getCodeState( std::string& state_out, int& status_code, std::string& status_message)
{
    status_code = 0;

    status_message = "INITIALISATION: OK";
    state_out = std::to_string(code_state);

    std::cout << "=======================================================" << std::endl;
    std::cout << "Basic CPP: GET STATE called" << std::endl;
    std::cout << "STATE is : " << state_out << std::endl;
    std::cout << "=======================================================" << std::endl;
}

// =======================================
//             SET STATE
// =======================================
void restoreCodeState( std::string state, int& status_code, std::string& status_message)
{
    status_code = 0;
    status_message = "FINALISATION: OK";

    code_state = std::stoi( state );
    std::cout << "=======================================================" << std::endl;
    std::cout << "Basic CPP: RESTORE STATE called" << std::endl;
    std::cout << "STATE TO BE RESTORED : " << code_state << std::endl;
    std::cout << "=======================================================" << std::endl;
}
// ================================================================================================================
//                                           TIMESTAMP SUBROUTINES
// ================================================================================================================

// =======================================
//             GET TIMESTAMP
// =======================================
void getTimestampCpp(double& timestamp_out, int& status_code, std::string& status_message)
{
    timestamp_out = timestamp;
}

// ================================================================================================================
//                                           INIT SUBROUTINES
// ================================================================================================================

// =======================================
//             INITIALISATION
// =======================================

void initCode(int& status_code, std::string& status_message)
{
    std::string parameters_str;
    initCode(parameters_str, status_code, status_message);
}

// =======================================
//             INITIALISATION
// =======================================

void initCode(std::string parameters_str, int& status_code, std::string& status_message)
{
    IdsNs::IDS::core_profiles core_profiles_in;
    IdsNs::IDS::distribution_sources distribution_sources;
    initCode(core_profiles_in, distribution_sources, parameters_str, status_code, status_message);
}

// =======================================
//        LEGACY-XML INITIALISATION
// =======================================

void initCode(const IdsNs::IDS::core_profiles& core_profiles_in,
                          IdsNs::IDS::distribution_sources& distribution_sources_out,
                          IdsNs::codeparam_t xml_params,
                          int& status_code, std::string& status_message)
{
    std::string parameters_str(*(xml_params.parameters));
    initCode(core_profiles_in, distribution_sources_out, parameters_str, status_code, status_message);
}

// =======================================
//        LEGACY-XML INITIALISATION
// =======================================

void initCode(IdsNs::codeparam_t codeparam, int& status_code, std::string& status_message)
{
    IdsNs::IDS::core_profiles core_profiles_in;
    IdsNs::IDS::distribution_sources distribution_sources;
    initCode(core_profiles_in, distribution_sources, codeparam, status_code, status_message);
}

// =======================================
//             INITIALISATION
// =======================================

void initCode(const IdsNs::IDS::core_profiles& core_profiles_in,
                          IdsNs::IDS::distribution_sources& distribution_sources_out,
                          int& status_code, std::string& status_message)
{
    IdsNs::codeparam_t codeparam;
    initCode(core_profiles_in, distribution_sources_out, codeparam, status_code, status_message);
}

// =======================================
//             INITIALISATION
// =======================================

void initCode(const IdsNs::IDS::core_profiles& core_profiles_in,
                          IdsNs::IDS::distribution_sources& distribution_sources_out,
                          std::string parameters_str,
                          int& status_code, std::string& status_message)
{
    int idsSize = -1;
    status_code = 0;

    status_message = "Basic methods CPP: INIT: OK";

    printf("=======================================================\n");
    printf("Basic methods CPP: INIT starts\n");
    printf("-------------------------------------------------------\n");

    idsSize = core_profiles_in.time.extent(0);
    printf("Size of input IDS  = %d\n", idsSize);

    distribution_sources_out.time.resize(idsSize);

    // Fill in the output IDS (Physical data)
    for(int i=0; i < idsSize; i++)
    {
        // Time : copy from input IDS
        distribution_sources_out.time(i) = core_profiles_in.time(i);
    }

    distribution_sources_out.ids_properties.homogeneous_time = 1;
    distribution_sources_out.code.name   = "basic_methods_cpp";
    distribution_sources_out.code.version   = "1.0";

    printf("-------------------------------------------------------\n");
    printf("Basic methods CPP: INIT ends\n");
    printf("=======================================================\n");
}

// ================================================================================================================
//                                           FINALIZE SUBROUTINES
// ================================================================================================================

// =======================================
//             FINALISATION
// =======================================
void cleanUp(int& status_code, std::string& status_message)
{
    IdsNs::IDS::distribution_sources distribution_sources_in;
    IdsNs::IDS::core_profiles core_profiles_out;
    cleanUp(distribution_sources_in, core_profiles_out, status_code, status_message);
}

// =======================================
//             FINALISATION
// =======================================
void cleanUp(const IdsNs::IDS::distribution_sources& distribution_sources_in,
                        IdsNs::IDS::core_profiles& core_profiles_out,
                        std::string parameters_str,
                        int& status_code, std::string& status_message)
{
    int idsSize = -1;

    status_code = 0;
    status_message = "FINALISATION: OK";
    printf("=======================================================\n");
    printf("Basic methods CPP: FINALISATION starts\n");
    printf("-------------------------------------------------------\n");

    idsSize = distribution_sources_in.time.extent(0);
    printf("Size of input IDS  = %d\n", idsSize);

    core_profiles_out.time.resize(idsSize);

    // Fill in the output IDS (Physical data)
    for(int i=0; i < idsSize; i++)
    {
        // Time : copy from input IDS
        core_profiles_out.time(i) = distribution_sources_in.time(i);
    }

    core_profiles_out.ids_properties.homogeneous_time = 1;
    core_profiles_out.code.name   = "basic_methods_cpp";
    core_profiles_out.code.version   = "1.0";

    printf("-------------------------------------------------------\n");
    printf("Basic methods CPP: FINALISATION ends\n");
    printf("=======================================================\n");
}

void cleanUp(std::string parameters_str, int& status_code, std::string& status_message)
{
    IdsNs::IDS::distribution_sources distribution_sources_in;
    IdsNs::IDS::core_profiles core_profiles_out;

    cleanUp(distribution_sources_in, core_profiles_out, parameters_str, status_code, status_message);
}

void cleanUp(const IdsNs::IDS::distribution_sources& distribution_sources_in,
                        IdsNs::IDS::core_profiles& core_profiles_out,
                        int& status_code, std::string& status_message)
{
    std::string parameters_str = "";
    cleanUp(distribution_sources_in, core_profiles_out, parameters_str, status_code, status_message);
}

// =======================================
//      LEGACY-XML FINALIZE FUNCTIONS
// =======================================

void cleanUp(IdsNs::codeparam_t codeparam, int& status_code, std::string& status_message)
{
    IdsNs::IDS::distribution_sources distribution_sources_in;
    IdsNs::IDS::core_profiles core_profiles_out;

    cleanUp(distribution_sources_in, core_profiles_out, codeparam, status_code, status_message);
}

void cleanUp(const IdsNs::IDS::distribution_sources& distribution_sources_in,
                        IdsNs::IDS::core_profiles& core_profiles_out,
                        IdsNs::codeparam_t codeparam,
                        int& status_code, std::string& status_message)
{
    std::string parameters_str(*(codeparam.parameters));
    cleanUp(distribution_sources_in, core_profiles_out, parameters_str, status_code, status_message);
}

// ================================================================================================================
//                                           MAIN SUBROUTINES
// ================================================================================================================

// =======================================
//             MAIN
// =======================================
void codeStep(int& status_code, std::string& status_message)
{
    std::string parameters_str;
    codeStep(parameters_str, status_code, status_message);
}

// =======================================
//             MAIN
// =======================================
void codeStep(std::string parameters_str, int& status_code, std::string& status_message)
{
    IdsNs::IDS::equilibrium equilibrium_in;
    IdsNs::IDS::equilibrium equilibrium_out;
    codeStep(equilibrium_in, equilibrium_out, parameters_str, status_code, status_message);
}

// =======================================
//             MAIN
// =======================================
void codeStep(const IdsNs::IDS::equilibrium& equilibrium_in, IdsNs::IDS::equilibrium& equilibrium_out,
                     int& status_code, std::string& status_message)
{
    std::string parameters_str;
    codeStep(equilibrium_in, equilibrium_out, parameters_str, status_code, status_message);
}

// =======================================
//             MAIN
// =======================================
void codeStep(const IdsNs::IDS::equilibrium& equilibrium_in, IdsNs::IDS::equilibrium& equilibrium_out,
                     std::string parameters_str,
                     int& status_code, std::string& status_message)
{
    int idsSize = -1;
    int idsTimeMode = IDS_TIME_MODE_UNKNOWN;

    // INITIALISATION OF ERROR FLAG
    status_code = 0;

    // INITIAL DISPLAY

    printf("=======================================================\n");
    printf("Basic methods CPP: MAIN starts\n");
    printf("-------------------------------------------------------\n");

    // CHECK IF INPUT IDS IS VALID
    idsSize = equilibrium_in.time.extent(0);
    idsTimeMode = equilibrium_in.ids_properties.homogeneous_time;
    if ( idsTimeMode != IDS_TIME_MODE_HOMOGENEOUS && idsSize > 0)
    {
        // ERROR IF THE CODE DOES NOT COMPLETE TO THE END
        status_code = -1;
        status_message = "Error in basic_methods_cpp: input IDS not valid";
        return;
    }

    // READ PARAMETERS
    auto parameters_tuple = extract_values(parameters_str);
    int ntimes = std::get<0>(parameters_tuple);
    float multiplication_factor = std::get<1>(parameters_tuple);

    // MANDATORY FLAG (UNIFORM TIME HERE)
    equilibrium_out.ids_properties.homogeneous_time = IDS_TIME_MODE_HOMOGENEOUS;
    equilibrium_out.time.resize(idsSize);
    equilibrium_out.code.output_flag.resize(idsSize);

    // Fill in the output IDS (Physical data)
    for(int i=0; i < idsSize; i++)
    {
        // Time : copy from input IDS
        equilibrium_out.time(i) = equilibrium_in.time(i);
        // MODIFY OUTPUT FLAG
        equilibrium_out.code.output_flag(i) = equilibrium_in.time(i) * ntimes * multiplication_factor * code_state;
    }

    // FILL IDS INFO
    equilibrium_out.ids_properties.homogeneous_time = 1;
    equilibrium_out.code.name   = "basic_methods_cpp";
    equilibrium_out.code.version   = "1.0";

    printf("Size of input IDS  = %d\n", idsSize);

    // INITIALISATION OF STATUS INFO
    status_message = "Status info of basic_methods_cpp CPP";
    code_state++;
    timestamp+=10;

    // FINAL DISPLAY
    printf("-------------------------------------------------------\n");
    printf("Basic methods CPP: MAIN ends\n");
    printf("=======================================================\n");
}

// =======================================
//            LEGACY-XML MAIN
// =======================================
void codeStep(IdsNs::codeparam_t xml_params, int& status_code, std::string& status_message)
{
    IdsNs::IDS::equilibrium equilibrium_in;
    IdsNs::IDS::equilibrium equilibrium_out;
    codeStep(equilibrium_in, equilibrium_out, xml_params, status_code, status_message);
}


// =======================================
//            LEGACY-XML MAIN
// =======================================
void codeStep(const IdsNs::IDS::equilibrium& equilibrium_in, IdsNs::IDS::equilibrium& equilibrium_out,
                     IdsNs::codeparam_t xml_params,
                     int& status_code, std::string& status_message)
{
    std::string parameters_str(*(xml_params.parameters));
    codeStep(equilibrium_in, equilibrium_out, parameters_str, status_code, status_message);
}
